src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1336 @@
+      _bsonDSL_restoreWarnings ();                     \
+   } while (0)
+
+/**
+ * @brief Build a BSON document by appending to an existing bson_t document
+ *
+ * @param Pointer The document upon which to append
+ * @param ... The Document elements to append to the document
+ */
+#define bsonBuildAppend(BSON, ...) \
+   _bsonDSL_eval (_bsonBuildAppend (BSON, __VA_ARGS__))
+#define _bsonBuildAppend(BSON, ...)                                    \
+   _bsonDSL_begin ("Appending to document '%s'", _bsonDSL_str (BSON)); \
+   _bsonDSL_disableWarnings ();                                        \
+   /* Save the dsl context */                                          \
+   struct _bsonBuildContext_t _bbCtx = {                               \
+      .doc = &(BSON),                                                  \
+      .parent = _bsonBuildContextThreadLocalPtr,                       \
+   };                                                                  \
+   /* Reset the context */                                             \
+   _bsonBuildContextThreadLocalPtr = &_bbCtx;                          \
+   bsonBuildError = NULL;                                              \
+   _bsonBuildAppendWithCurrentContext (__VA_ARGS__);                   \
+   /* Restore the dsl context */                                       \
+   _bsonBuildContextThreadLocalPtr = _bbCtx.parent;                    \
+   _bsonDSL_restoreWarnings ();                                        \
+   _bsonDSL_end
+
+/**
+ * @brief Build a new BSON document and assign the value into the given
+ * pointer.
+ */
+#define bsonBuild(BSON, ...)                                              \
+   _bsonDSL_begin ("Build a new document for '%s'", _bsonDSL_str (BSON)); \
+   bson_t *_bbDest = &(BSON);                                             \
+   bson_init (_bbDest);                                                   \
+   bsonBuildAppend (*_bbDest, __VA_ARGS__);                               \
+   _bsonDSL_end
+
+/**
+ * @brief Declare a variable and build it with the BSON DSL @see bsonBuild
+ */
+#define bsonBuildDecl(Variable, ...)   \
+   bson_t Variable = BSON_INITIALIZER; \
+   bsonBuild (Variable, __VA_ARGS__)
+
+
+struct _bsonBuildContext_t {
+   /// The document that is being built
+   bson_t *doc;
+   /// The key that is pending an append
+   const char *key;
+   /// The length of the string given in 'key'
+   int key_len;
+   /// The index of the array being built (if applicable)
+   int index;
+   /// A buffer for formatting key strings
+   char index_key_str[16];
+   /// The parent context (if building a sub-document)
+   struct _bsonBuildContext_t *parent;
+};
+
+/// A pointer to the current thread's bsonBuild context
+_bson_thread_local_comdat struct _bsonBuildContext_t
+   *_bsonBuildContextThreadLocalPtr = NULL;
+
+struct _bsonVisitContext_t {
+   const bson_t *doc;
+   bson_iter_t iter;
+   const struct _bsonVisitContext_t *parent;
+   int index;
```

Comment:
```suggestion
   size_t index;
```
`_bsonVisitContext_t::index` should have type `size_t`.